app/controllers/concerns/cybersource.rb

```diff
@@ -0,0 +1,192 @@
+
+  class PaymentFailed < StandardError; end
+  class InvalidSignature < StandardError; end
+
+  included do
+    def cybersource_request
+      PaymentRequest.new(**params.permit(:user, :amount, :session_id).to_h.symbolize_keys).sign!
+    end
+
+    def cybersource_response
+      PaymentResponse.new(params.permit!.to_h).validate!
+    end
+  end
+
+  # Methods for encoding and decoding transaction data
+  class Security
+    class << self
+      # Cybersource uses a base64 encoded HMAC SHA256 digest to sign the data
+      def generate_signature(data)
+        payload = data.map { |key, value| "#{key}=#{value}" }.join(',')
+        Base64.encode64(OpenSSL::HMAC.digest('sha256', secret_key, payload)).strip
+      end
+
+      # If the signature matches the data, we know it hasn't been tampered with
+      def valid_signature?(signature, data)
+        generate_signature(data) == signature
+      end
+
+      def secret_key
+        Settings.cybersource.secret_key
+      end
+    end
+  end
+
+  # Request sent to Cybersource to initiate an external checkout
+  class PaymentRequest
+    include Enumerable
+
+    attr_reader :user, :amount, :session_id, :signature, :signed_date_time
+
+    # Set of fields we use to generate the signature and that Cybersource verifies
+    REQUEST_SIGNED_FIELDS = %i[access_key profile_id transaction_uuid signed_date_time
+                               locale transaction_type reference_number amount currency unsigned_field_names
+                               signed_field_names merchant_defined_data1].freeze
+
+    def initialize(user:, amount:, session_id:)
+      @user = user
+      @amount = amount
+      @session_id = session_id
+      @signed_fields = REQUEST_SIGNED_FIELDS
+      @unsigned_fields = []
+      @signed_date_time, @signature = nil
+    end
+
+    # Hash form including all data and signature
+    def to_h
+      signed_data.merge(unsigned_data).merge(signature: @signature)
+    end
+
+    # Support iterating keys and values for rendering as hidden form fields
+    def each(&block)
+      to_h.each(&block)
+    end
+
+    # Generate a security signature for the data and add a timestamp
+    def sign!
+      @signed_date_time = Time.now.utc.strftime('%Y-%m-%dT%H:%M:%SZ')
+      @signature = Security.generate_signature(signed_data)
+
+      self
+    end
+
+    def validate!
+      raise InvalidSignature unless valid?
+
+      self
+    end
+
+    def signed?
+      @signature.present?
+    end
+
+    def valid?
+      Security.valid_signature?(@signature, to_h)
+    end
+
+    private
+
+    # Comma-separated parameter version of signed_fields
+    def signed_field_names
+      @signed_fields.join(',')
+    end
+
+    # Comma-separated parameter version of unsigned_fields
+    def unsigned_field_names
+      @unsigned_fields.join(',')
+    end
+
+    # Hash of signed fields and their values
+    def signed_data
+      @signed_fields.index_with { |field| send(field) }
```

Comment:
Should this be memoized?  If you call `obj.sign!; obj.to_h` it's called two times. I'd think after signing, we wouldn't want to produce this value again.

Reply:
yeah, I was holding off memoizing until I finished testing, but this seems like a good candidate. will do

Reply:
memoized